packages/core/bootstrap/src/lib/ws/reducer.ts

```diff
@@ -1,65 +1,65 @@
+    }
+  }
 }
 
-const initConnectionsState: ConnectionsState = { active: {}, connecting: {}, wasEverConnected: {}, total: 0 }
-
-export const connectionsReducer = createReducer<ConnectionsState>(
-  initConnectionsState,
-  (builder) => {
-    builder.addCase(actions.connected, (state, action) => {
-      // Add connection
-      const { key } = action.payload.config.connectionInfo
-      state.active[key] = action.payload.config
-      state.connecting[key] = 0
-      state.wasEverConnected[key] = true
-      // Increment num of active connections
-      state.total++
-    })
-
-    builder.addCase(actions.connect, (state, action) => {
-      const { key } = action.payload.config.connectionInfo
-      const isActive = !!state.active[key]
-      if (isActive) return
-
-      const isConnecting = !isNaN(Number(state.connecting[key]))
-      state.connecting[key] = isConnecting ? state.connecting[key] + 1 : 1
-    })
-
-    builder.addCase(actions.connectionError, (state, action) => {
-      state.connecting[action.payload.connectionInfo.key] = 0
-      delete state.active[action.payload.connectionInfo.key]
-    })
-
-    builder.addCase(actions.disconnected, (state, action) => {
-      // Remove connection
-      const { key } = action.payload.config.connectionInfo
-      delete state.active[key]
-      state.connecting[key] = 0 // turn off connecting
-      // Decrement num of active connections
-      if (state.wasEverConnected[key]) state.total--
-    })
-  },
-)
+const initConnectionsState: ConnectionsState = { total: 0, all: {} }
+
+export const connectionsReducer = createReducer<ConnectionsState>(initConnectionsState, builder => {
+  builder.addCase(actions.connected, (state, action) => {
+    // Add connection
+    const { key } = action.payload.config.connectionInfo
+    state.all[key] = {
+      active: true,
+      connecting: 0,
+      wasEverConnected: true,
+    }
+  })
+
+  builder.addCase(actions.connect, (state, action) => {
+    const { key } = action.payload.config.connectionInfo
+    const isActive = state.all[key]?.active
+    if (isActive) return
+
+    const isConnecting = !isNaN(Number(state.all[key]?.connecting))
+    state.all[key] = {
+      active: false,
+      connecting: isConnecting ? state.all[key].connecting + 1 : 1,
+    }
+  })
+
+  builder.addCase(actions.connectionError, (state, action) => {
+    state.all[action.payload.connectionInfo.key].connecting = 0
+    state.all[action.payload.connectionInfo.key].active = false
+  })
+
+  builder.addCase(actions.disconnected, (state, action) => {
+    // Remove connection
+    const { key } = action.payload.config.connectionInfo
+    state.all[key].active = false
+    state.all[key].connecting = 0 // turn off connecting
+  })
+
+  builder.addMatcher(
+    isAnyOf(actions.connect, actions.connected, actions.connectionError, actions.disconnected),
+    state => {
+      state.total = Object.values(state.all).filter(s => s?.active).length
+    },
+  )
+})
```

Comment:
Refactored the state, and added an additional reducer that is triggered after any of the reducer events (reducers are used in the order they are listed) to calculate the total connections.

If this looks good, I can also refactor the subscriptions state to be similar

Reply:
This looks good to me